File: binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
  int ans = Integer.MIN_VALUE;
  public int maxPathSum(TreeNode root) {
    dfs(root);
    return ans;
  }
  int dfs(TreeNode root) {
    if (root == null) return 0;
    int left_ans = 0, right_ans = 0;
    left_ans = Math.max(dfs(root.left), left_ans);
    right_ans = Math.max(dfs(root.right), right_ans);
    int ans_node = left_ans + right_ans + root.val;
    ans = Math.max(ans, ans_node);
    return Math.max(left_ans + root.val, right_ans + root.val);
  }
}

File: binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
        private:
        int dfs(TreeNode* root,int &ans){
                if(root==NULL) return 0;
                int left_ans=0,right_ans=0;
                left_ans=max(left_ans,dfs(root->left,ans));
                right_ans=max(right_ans,dfs(root->right,ans));
                int ans_here=left_ans+right_ans+root->val;
                ans=max(ans_here,ans);
                return max(left_ans+root->val,root->val+right_ans);
        }
        public:
    int maxPathSum(TreeNode* root) {
        // you can take either the left or the right nodes
            int ans=INT_MIN;
            
            dfs(root,ans);
            return ans;
    }
};